### DISCUSSION TITLE
Reviews

### DISCUSSION THREAD

**Discussion by JeremyLARDENOIS** [Pokemon-PythonRed/Pokemon-PythonRed#34] [General]
Hello, do you remember me ?

I saw that you made a lot of change and I decided to test it ;)
I tested it on Windows 10 and Linux and this is my review.
All what I will say are just advices and you can choose to take it or not, or discuss about them 🙂

### On windows:
- For the installation, i don't think that `git` must be mandatory to play to PPR. I think that you need to diffency the case where a player want play and the case where someone wants contribute. By exemple, the person can download the project in zip.
-  You have a lot of requirements that you don't use. Players can plays without install (Flask, pygame). Do you plan to use it soon ?

### On linux:
- the lib `msvcrt` doesn't exist on linux, so for testing, i commented all reference to `g()`
- to install, i used `pip3` command but you can use `python3 -m pip` instead. i don't know if it's a mistake but you forgot the word install...

## Code and game
- It could be cool if we can change `textSpeed` in game
- In the same way, catch keyboard input to print all the text in one time
- use classes instead of dict. it's a more convenent way and help to avoid some crashes
- Save `save` into a file to keep data
- seperate code into multiple files
 
Also, as you know i have a fork of your project that is Linux supported. Do you prefer that i keep doing on my side or that i try to do a version Windows and Linux in your project ?

That's all for bad things. As I already said, i love the idea and i'm curious to see the results because i was pokemon version red player. If you need help in anything, don't hesitate to ask me !

Happy to help,
Jerem

**Comment by riverpaper** [comment]
Yes, I remember you! Thanks for staying with this project; I don't have access to other operating systems, so I appreciate the cross-platform testing you do. Here's what I have to say about each of your points:

### (Windows)
- I think I understand what you're trying to say:
> If a person wants to contribute, then they need to clone the repository.
> If they only want to play the game, they can just download the `.zip` file.

You are correct in saying this, I'll add that to the README soon.
- The program is only importing the libraries it uses. The others, such as `pygame`, are commented out until I need them.
### (Linux)
- If the library doesn't exist, then that's a problem. From a quick Google search, I found out that there's a Linux library called `getch` that can do the same thing.
- You're right, I forgot the word `install` by mistake.
### (Code)
- I'm planning to add a feature that lets you change the text speed from the Options menu, just like the original games! Since there's not much dialogue in the game at this point, I figured it doesn't matter just yet, but I'm planning to add it.
- I'm not sure what you're trying to say here, sorry.
- I'm planning to use classes for a few things, like storing Pokemon in the party and boxes. I think that dictionaries also work well for things like storing items, as well as being easier to use. Also, I haven't used classes very much, so I feel more confident with dictionaries.
- Saving is already implemented; I thought it would be difficult, so I made it one of the first things I handled. You can use the `m` command from the overworld (Player's House, etc.), and it will open the menu. From there, you can choose `[s] - Save`, and your data will be saved to `save.json` in the same directory as `main.py`. On the next restart, use `[1] - Continue` on the start menu and it should restore your data.
- Is it better to have the code in multiple files? I've done a bit of that already, keeping long dictionaries inside `.json` files instead of directly in `main.py`. If keeping code separate is a common practice, just leave a Pull Request and I'll be happy to accept!
### (Misc.)
- I would prefer to have all the cross-platform functions inside one program since it should be entirely possible with Python's `platform` library.
- Thanks for all your help! I do have one request: could you please test the `getch` library on Linux and tell me the code I need to get the same effect as `msvcrt.getch()`? Then I could add support for both platforms.

Thanks again, have a good one!

**Comment by JeremyLARDENOIS** [comment]
Your welcome! :wink:

> The program is only importing the libraries it uses. The others, such as pygame, are commented out until I need them.

Yes, it's true, but you ask to do a pip install whereas it's not needed for the moment. If you think that these libraries will be required soon, you can let it.

P.S.: Let it, we need it for getch library :slightly_smiling_face: 

> In the same way, catch keyboard input to print all the text in one time

In pokemon, when the text is printing, you can accelerate the printing by pressing a button. I think it could be cool to implement it. In a quick search, I found a library `keyboard` https://pypi.org/project/keyboard/ . I don't know if it's the right way to do this.

### About classes :

In fact, you can continue to use dict if you are more familiar with it. The mainly advantage with classes is that you can encapsulate function and data in the same structure. Also, I think that it's easier to have errors management with classes: in app/main.py, in line 406, you are trying to access to `save["trainer"]["name"]` whereas `save["trainer"]̀ is not defined, and that makes the program crash. (I will open an issue about this). But like I said, if you are more familiar with dict, this is not a problem.

>  Saving is already implemented; I thought it would be difficult, so I made it one of the first things I handled. You can use the m command from the overworld (Player's House, etc.), and it will open the menu. From there, you can choose [s] - Save, and your data will be saved to save.json in the same directory as main.py. On the next restart, use [1] - Continue on the start menu and it should restore your data.

Oh ok! my bad

> Is it better to have the code in multiple files? I've done a bit of that already, keeping long dictionaries inside .json files instead of directly in main.py. If keeping code separate is a common practice, just leave a Pull Request and I'll be happy to accept!

It depends about the size of your project. It's just a question of organization and there is a lot of git advantage when the project grows up.
You can read an article here about this <https://www.teclado.com/30-days-of-python/python-30-day-21-multiple-files>

Let me know if you have some questions about this, or if you disagree (it's possible, i have no problem with that because we learn things everydays :wink: )

Some issues and PR are comming :slightly_smiling_face:

**Comment by riverpaper** [comment]
> In pokemon, when the text is printing, you can accelerate the printing by pressing a button. I think it could be cool to implement it. In a quick search, I found a library keyboard https://pypi.org/project/keyboard/. I don't know if it's the right way to do this.

That's a cool idea, I hadn't thought of that... If I find a way, I'll implement it.

> It depends on the size of your project. It's just a question of organization and there is a lot of git advantage when the project grows up. You can read an article here about this https://www.teclado.com/30-days-of-python/python-30-day-21-multiple-files

That makes sense, I'll keep it in mind.
